src/ucp/tag/tag_send.c

```diff
@@ -131,13 +132,17 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_send_nb,
     ucp_request_t *req;
     size_t length;
     ucs_status_ptr_t ret;
+    uct_memory_type_t mem_type;
 
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
     ucs_trace_req("send_nb buffer %p count %zu tag %"PRIx64" to %s cb %p",
                   buffer, count, tag, ucp_ep_peer_name(ep), cb);
 
-    if (ucs_likely(UCP_DT_IS_CONTIG(datatype))) {
+    /* Detect memory type */
+    ucp_memory_type_detect_mds(ep->worker->context, (void *)buffer, count, &mem_type);
```

Comment:
this will hurt small message performance, need to detect memory only if this is not short
and if cuda is active, max_short could be 0